README.md
04/11/22:
- đã đổi cấu trúc project, bỏ module

TTC-CAM consumer tạo ngày 06/01/2022
TTC-CAM consumer có 2 instance :
- ctcd : chạy service, kết nối theo tài khoản ctc, topic=ctc/ctc1/#
- ttc-cam : chạy console để test, kết nối theo tài khoản ttc/xuanquang, topic=ttc/ttc1/#
2 instance chỉ khác nhau file cấu hình, trong code giống nhau hết


CHANGELOG.md
# Changelog
// timeStamp, lat, lng, speed, acc, sdStatus, fuel, angle, distance, driver_name, driver_license, isOverSpeed, drivingElapse, isSessionOvertime, dayStretch
11/03/23:
- tách riêng ra thành 1 method getMyGeo
- trước đây các case QCVN không được xử lý position, bi giờ cho getMyGeo hết : updateQCVNOverSpeed R, updateQCVNPark R lng-lat, updateQCVNDriving R lng-lat, updateMileage R

07/03/23:
- bảng cam trong rts chưa dùng
- tạo thêm 1 bảng photo : photo_id, device_id, channel, lat, lng, speed, receive_time, path, enable, position
- khi subscribe qos=2

06/03/23:
- phân quyền cho folder deviceID và file ảnh : gọi shell chown cho tài khoản dtd

03/03/23:
- MqttMessage : dùng mảng byte rồi check để lấy string
- photo payload : 0xFF 0xD8 . . . . 0xFF 0xD9;chuỗi param : lấy khoảng 100 byte cuối cùng, tìm ký tự ';', sau ký tự ; là nội dung param
- param gồm : unixTime,lat,lng,speed,driver,license,frameSize, dài khoảng <100 byte
- sửa file config.xml : trong thẻ watchdog thêm photos_path

16/02/23:
- khi calib lỗi thì try catch -> vẫn xử lý location và fuel=-1

19/01/23:
- fw v1.6.1 đã add các feature về network
{"id":"ttc0_f412fa4313b8","time":1674101346,"type":"network","phone_number":"84369370621","operator":"Viettel","rssi":19,"imsi":"452048825588958","iccid":"89840480088255889583"}

04/01/23:
- elapse bị sai trong case : miss gps

31/12/22:
- vtrackd có sửa processPower

30/12/22:
- db : split_log_record_by_device : clear day_elapse trong device và moving_elapse trong device_log
- bỏ khái niệm movingElapse trong processPower, nếu update session thì cộng dồn vào cái cũ
- bỏ moving_elapse trong device_log
- trong 4 case : chỉ làm case moving->moving thì movingElaspe chỉ có ý nghĩa để cộng dồn vào day_elapse của device (vẫn phải cần ACC)
- khi lấy lastInsertRecord st.getGeneratedKeys() : có mã mới update, không có mã không được update, chỉ sửa vtrackd processPower, chạy được thì sửa trong ctcd -> đã sửa luôn ctcd
-> sửa luôn lastInsertID trong các method, nếu success thì quay lại sửa trong vtrackd (vtrackd phải giữ lastInsertId lại như cũ)
- vtrackd cho thêm finally đóng cs, st, con
- 
29/12/22:
- bảng power : khi update record phân ra 2 trường hợp : nếu ON thì update movingElapse, nếu OFF thì không
- day_elapse trong bảng device nó chỉ lưu gía trị movingElapse gần nhất -> đã thêm biến dayMovingElapse -> sau khi split không clear day_elapse

28/12/22:
- moving_elapse : tính thêm case moving->stop
- nếu qua ngày thì tạo record mới trong power
- có bug : khi tạo mới record power qua ngày thì nó vẫn giữ nguyên giá trị elapse

22/12/22:
- khôi phục lại day_elaspe trong db : chỉ tính khi moving to moving, 3 case khác chỉ lưu
- db đã tăng thêm 1 cột moving_elapse trong bảng power, đã add cho rts và rtslog
- summary_report tái sử dụng cột power_off_count làm moving_elapse
- sửa lại algorithm : trong device_log chỉ lưu đúng moving_elapse chứ không lưu cả ngày, chỉ có device mới lưu moving cả ngày
case bắt đầu moving : cho moving_elapse=0
case đang moving : cộng dồn như hiện tại
case moving -> stop : update moving_elapse vào device : không làm
- chưa xử lý được bảng power cột moving_elapse : làm sao để cộng dồn các record khi moving->R
- còn : khi bắt đầu move không tính moving_elapse -> xem lại
- khi có event : update version : đang comment

04/11/22:
- đã đổi cấu trúc project, bỏ module

18/10/22:
- đã fix summary_report
- bảng device_log : đổi tên cột driver_elapse -> day_stretch
- processLocation : 3 case (trừ StopToStop) đều phải update hoặc insert stretch vào cột day_stretch của device_log, làm song song với device

16/10/22:
- add fixGPS, dùng lại dayStretch
- khi validateIMEI : bỏ day_elapse

29/09/22:
- phiên bản ttc-cam chay thử với thư viện rabbitmq, không dùng paho : vì pahoo sau 1 thời gian thì không nhận data
- thay dần các listener và method của paho

20/09/22:
- fuel calib : tạo bảng calib trong DB
- device có thêm cột tank_height(int), calib(boolean) : check trước calib, nếu true thì bỏ qua tank height, nếu không cần calib thì căn cứ tank_height
- sửa các funtion liên quan tính summary : calculate_oil_report, . . .
- bảng device_log : cũng sửa tiền tố oil_
- khi query thông device device : lấy clientTaxCode của tài khoản quản lý (taxCode của đơn vị vận tải)
lưu value taxCode vào Tracker
- trong bảng slave_account, master_account db đổi tên cột mst thành tax_code
- file cấu hình : đổi tên thuộc tính mst -> taxcode

31/08/22:
- tạo thêm cột amqp_success trong bảng mileage, có hoán đổi vị trí cột, check lại ctcd
- vì gởi amqp theo bảng mileage nên tạo thêm mileageID trong thuộc tính tracker
- kiểm tra lại vtrackd
- đã test gởi amqp đến DRVN tốt

30/08/22:
- tích hợp 2 lib : protobuf-java.jar và rabbitmq-client.jar
- gởi AMQP đến server

26/08/22:
- có bug : device gởi không có distance, arr[8] là distance, server phải tự tính

23/08/22:
- chuyển isLegalTime() vào trong Tracker
- khi offline thi isLegalTime=false -> elapse=0

16/08/22 :
- tạo overspeed violation
- tạo bảng overspeed_violation : os_violation_id
- trong class Tracker có 2 biến khác nhau : speedLimit, overSpeed và overSpeedViolation. overSpeed do device tính toán và gởi đến server, speedLimit do server lưu sẵn, overSpeedViolation là server tính toán vi phạm theo TT09
- sửa lại ý nghĩa elapse : là khoảng time lần cuối cùng gởi msg và time của lần hiện tại, tức là end_time trong DB và getTime của tracker. Với trường hợp OFFLINE MSG, elapse sẽ rối tung, nên khi đó = 0
- khi giả lập speed thì device không tính được overspeed, nên server sẽ nhận được false
- cột driving_id trong bảng device sẽ bỏ
- có 2 trường hợp sẽ xóa biến speed_vio_id trong bảng device : khi stop hoặc khi không bị violation, record trong bảng speed_violation đã tự tính realtime rồi, không cần cho end
-> chỉ có 1 case MovingToMoving là xử lý, còn 3 case kia kiểm tra speed_vio_id trong device, nếu > 0 thì clear=0
- bổ sung thêm duration để khác biệt với elapse : khi stop thì duration tính từ beginTime đến hiện tại của tracker

30/07/22:

src/main/java/dxq/Service.java
package dxq;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.ConnectionFactory;
import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;

import java.io.File;
import java.util.*;

import dxq.util.*;
import dxq.tracker.*;

public class Service implements MqttCallbackExtended{
    public static final Logger serviceLog = LogManager.getLogger(Service.class);
    private MyUtil util = null;
    private MemoryPersistence persistence = new MemoryPersistence();
    private MqttClient client;
    private MqttConnectOptions opt;
    private Channel channel;
    private TrackerAdapter adapter = null;
//    private static final String LOCATION_MSG = "gps";
//    private static final String FIRMWARE_MSG = "fw";
//    private static final String EVENT_MSG = "event";
//    private static final String STATUS_MSG = "status";
    String supplier1 = "ctc";
    String supplier2 = "ttc";
    String revision1 = "ctc1";
    String revision2 = "ctc2";
    String revision3 = "ttc1";
    String revision4 = "ttc2";

    public Service(){
        util = new MyUtil();
        opt = new MqttConnectOptions();
        opt.setCleanSession(true);
        opt.setUserName(util.getBrokerUsername());
        opt.setPassword(util.getBrokerPassword().toCharArray());
        opt.setAutomaticReconnect(true);
        opt.setMqttVersion(MqttConnectOptions.MQTT_VERSION_3_1_1);
        connectMqttCloud();
        connectAmqpDRVN(util);

        // test phan quyen folder
//        System.out.println("is linux="+MyUtil.isLinux());
//        String folderName = util.getPhotosPath() + "/test";
//        File folder = new File(folderName);
//        if (!folder.exists()) {
//            // neu chua co folder thi create folder
//            boolean success = folder.mkdir();
//            System.out.println("Create folder " + folderName + " success : " + success);
//            if (!success) {
//                // neu tao folder khong duoc thi return
////                System.out.println("Create folder error, imei=" + imei);
//                return;
//            }
//            // phan quyen folder
//            if(MyUtil.isLinux()){
//                try{
//                    Runtime.getRuntime().exec("chown " + MyUtil.chown + " " + folderName);
//                }catch(Exception ex){ex.printStackTrace();}
//            }else{
//            }
//        }
//        int minLevel = 0;
//        int maxLevel = 700;
//        int minValue = 5;
//        int maxValue = 60;
//        float sampleValue = (float)(maxValue - minValue)/(maxLevel - minLevel);
//        System.out.println("float="+sampleValue);
    }

    private void connectMqttCloud() {
        serviceLog.info("Connecting mqtt cloud "+util.getBrokerAddress()+" . . .","info");
        try {
            client = new MqttClient(util.getBrokerAddress(), util.getClientId(), persistence);
            client.setCallback(this);
            client.connect(opt);
        }catch(MqttException ex) {
//            serviceLog.error(ex.toString(),"error");
            ex.printStackTrace();
        }
    }

    //--------------------------------------------------
    // subscribe
    public void subscribeCTCDevices() {
        String topic = util.getCtcTopic();
        try {
            if(client.isConnected()) {
                client.subscribe(topic, 2);
                serviceLog.info("subscribe : "+topic,"info");
                System.out.println("Subscribe CTC topic");
            }else{
                System.out.println("Connection be lost when subscribe");
            }
        }catch(Exception ex) {
            serviceLog.error(ex.toString(),"error");
            ex.printStackTrace();
        }
    }
    //--------------------------------------------------
    // implements MqttCakkbackExtended
    public void connectionLost(Throwable cause) {
        serviceLog.info("connectionLost","info");
        System.out.println("Connection be lost");
    }

    public void connectComplete(boolean complete,String uri) {
        System.out.println("connect="+complete+",uri="+uri);
        serviceLog.info("Connect to "+uri+" : "+complete,"info");
        subscribeCTCDevices();
//        if(complete) subscribeList(switchBeans.getSubscribeSwitchList());
    }

    public void deliveryComplete(IMqttDeliveryToken token) {
        serviceLog.info("deliveryComplete","info");
    }

    public synchronized void messageArrived(String topic, MqttMessage message) throws Exception{
//        System.out.println(topic+"--"+message);
        String[] top = topic.split("/");
        if((supplier1.equals(top[0]) || supplier2.equals(top[0])) && (revision1.equals(top[1]) || revision2.equals(top[1]) || revision3.equals(top[1]) || revision4.equals(top[1]))){
            String imei = top[2];
            String msgType = top[3];
//            System.out.println(imei + "," + msgType);
//            byte[] b = message.getPayload();
//            System.out.println("payload length=" + b.length);
//            Thread th = new TrackerAdapter(imei, msgType, message.toString(), this.channel, this.util);
            Thread th = new TrackerAdapter(imei, msgType, message.getPayload(), this.channel, this.util);
            th.start();
        }else{
            System.out.println("Syntax not matched");
        }
    }

    private void connectAmqpDRVN(MyUtil util) {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(util.getAmqpHost());
        factory.setUsername(util.getAmqpUsername());
        factory.setPassword(util.getAmqpPassword());
        factory.setPort(util.getAmqpPort());
        factory.setVirtualHost(util.getAmqpVirtualHost());

        try{
            com.rabbitmq.client.Connection connection = factory.newConnection();
            this.channel = connection.createChannel();
            System.out.println("channel=" + channel);
        }catch(Exception ex){
            ex.printStackTrace();
        }
    }
}


src/main/java/dxq/tracker/Tracker.java
package dxq.tracker;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Set;
import java.util.TreeMap;

public class Tracker{
    public int UPLOAD_PACKAGE_TYPE = 0;
    public static final int UPLOAD_DATA_CODE = 1;
    public static final int UPLOAD_COMMAND_CODE = 2;
    public static final int UPLOAD_EVENT_CODE = 3;
    public static final int UPLOAD_ERROR_CODE = 0;
    public int DEVICE_STATUS_CASE;

    public static final int MIN_MOVING_SPEED = 5;
    public static final int CTCTRACKER = 1;
//    public static final int CONCOX = 2;
//    public static final int NASIA = 3;
    public static final int DRIVING_LEGAL_ELAPSE = 4*3600;

    protected long unixTime = 0;
    protected int deviceID = 0;
    protected String name = "";
    protected String imei = "";
    protected String type = "";
    protected float lat = 0;
    protected float lng = 0;
    protected float prevLat = 0;
    protected float prevLng = 0;
    protected int speed = 0;
    protected int prevSpeed = 0;
    protected int distance = 0;
    protected int angle = 0;
    protected boolean gps = true;
    protected boolean power = false;
    protected boolean prevPower = false;
    protected long prevBeginTime = 0;
    protected long prevEndTime = 0;
    protected int elapse = 0;
    protected boolean fixGPS = false;// Quang add 16/10/22
    //----------------------------------------------------------------------------------------------------
    protected String address = null;
    protected long logID = 0;
    //----------------------------------------------------------------------------------------------------
    // Q add 28/07/2021 : bổ sung field theo QCVN của DB rts
    protected boolean overSpeed = false;
    // Quang add 16/08/22
    protected boolean overSpeedViolation = false;
    protected String driverName = "";
    protected String driverLicense = "";
    // Quang add 30/05/22
    protected int drivingElapse = 0;
    protected boolean drivingOvertime = false;
    // Quang add 20/08/22
    protected long speedVioID = 0;
    //----------------------------------------------------------------------------------------------------
    // các peripherial : fuel, temp, aircon, truck, dvr của device_log
    protected int fuelRaw = 0;
    protected int prev_fuel_value = 0;
    protected long prevFuelTime = 0;
    protected int fuelValue = 0;
    protected int tempValue = 0;
    protected boolean over_temp = false;
    protected boolean aircon_value = false;
    protected boolean truck_value = false;
    // dvr chưa làm
    protected boolean sd = false;
    protected boolean di2 = false;
    // các field sau của bảng device, dùng để đối chiếu tinh toán
    protected long power_id = 0;
    protected boolean fuelEnable = false;
    protected int tank_capac = 0;
    protected int fuelDelta = 0;
    protected long fuelId = 0;
    protected boolean temp_enable = false;
    protected int temp_limit = 0;
    protected boolean aircon_enable = false;
    protected long aircon_id = 0;
    protected boolean truck_enable = false;
    protected long truck_id = 0;
    protected boolean dvr_enable = false;
    protected long dvr_id = -1;
    protected boolean cam1_enable = false;
    protected boolean cam2_enable = false;
    protected boolean cam3_enable = false;
    protected String photo1Path = "";
    protected String photo2Path = "";
    protected String photo3Path = "";
//    protected long door_id = -1;
    protected int day_stretch = 0;
//    protected int day_elapse = 0;
    protected boolean nd91 = true;
    protected int speed_limit = 70;
    protected long mileageID = 0;
//    protected long driving_id = -1;
    protected String taxCode = "";
    // Quang add 10/12/22
    protected TreeMap<Integer, Integer> calibMap = null;
    // Quang add 22/12/22
//    protected int movingElapse = 0;
    protected int dayMovingElapse = 0;
    protected String numberPlate = "";

    //----------------------------------------------------------------------------------------------------
    public Tracker(String type) {
        this.type = type;
    }
    //----------------------------------------------------------------------------------------------------
    public void setTime(long unixTime) {
        this.unixTime = unixTime;
    }
    public long getTime() {
        return this.unixTime;
    }
    public void setID(int deviceID) {
        this.deviceID = deviceID;
    }
    public int getID() {
        return this.deviceID;
    }
    public void setName(String name) {
        this.name = name;
    }
    public String getName() {
        return this.name;
    }
    public void setIMEI(String imei) {
        this.imei = imei;
    }
    public String getIMEI() {
        return this.imei;
    }
    public void setType(String type) {
        this.type = type;
    }
    public String getType() {
        return this.type;
    }
    public void setLat(float lat) {
        this.lat = lat;
    }
    public float getLat() {
        return this.lat;
    }
    public void setLng(float lng) {
        this.lng = lng;
    }
    public float getLng() {
        return this.lng;
    }
    public void setPrevLat(float prevLat) {
        this.prevLat = prevLat;
    }
    public float getPrevLat() {
        return this.prevLat;
    }
    public void setPrevLng(float prevLng) {
        this.prevLng = prevLng;
    }
    public float getPrevLng() {
        return this.prevLng;
    }
    public void setAngle(int angle) {
        this.angle = angle;
    }
    public int getAngle() {
        return this.angle;
    }
    public void setSpeed(int speed) {
        this.speed = speed;
    }
    public int getSpeed() {
        return this.speed;
    }
    public void setPrevSpeed(int prevSpeed) {
        this.prevSpeed = prevSpeed;
    }
    public int getPrevSpeed() {
        return this.prevSpeed;
    }

    public void setGps(boolean gps) {
        this.gps = gps;
    }
    public boolean isGps() {
        return this.gps;
    }
    public void setPower(boolean power) {
        this.power = power;
    }

    public boolean getPower() {
        return this.power;
    }

    public void setPrevPower(boolean prevPower) {
        this.prevPower = prevPower;
    }

    public boolean getPrevPower() {
        return this.prevPower;
    }

    public void setPrevBeginTime(long prevBeginTime) {
        this.prevBeginTime = prevBeginTime;
    }
    public long getDuration() {
        return this.unixTime - this.prevBeginTime;
    }

    public void setElapse(int elapse) {
        this.elapse = elapse;
    }
    public int getElapse() {
        return this.elapse;
    }
    // stretch=distance
    public void setDistance(int distance) {
        this.distance = distance;
    }
    public int getDistance() {
        return this.distance;
    }
    //----------------------------------------------------------------------------------------------------

    public void setAddress(String address) {
        if(address.equals("")) this.address= null;
        else{
            address = address.replace("'","");
            address = address.replace("\"","");
            this.address = address;
        }
    }
    public String getAddress() {
        return this.address;
    }
    public void setLogID(long logID) {
        this.logID = logID;
    }
    public long getLogID() {
        return this.logID;
    }
    //----------------------------------------------------------------------------------------------------

    // Q add 28/07 :
    public void setOverSpeed(boolean overSpeed){
        this.overSpeed = overSpeed;
    }

    public boolean isOverSpeed(){
        return this.overSpeed;
    }

    public void setDriverName(String driverName){
        this.driverName = driverName;
    }

    public String getDriverName(){
        return this.driverName;
    }

    public void setDriverLicense(String driverLicense){
        this.driverLicense = driverLicense;
    }

    public String getDriverLicense(){
        return this.driverLicense;
    }

    //----------------------------------------------------------------------------------------------------
    // Q add 28/07 : bổ sung fuel, . . .
    public void setFuelRaw(int fuelRaw){
        this.fuelRaw = fuelRaw;
    }

    public float getFuelRaw(){
        return this.fuelRaw;
    }

    public void setPrevFuel(int prevFuelValue){
        this.prev_fuel_value = prevFuelValue;
    }

    public int getPrevFuel(){
        return this.prev_fuel_value;
    }

    // Quang add 22/02/23
    public void setPrevFuelTime(long prevFuelTime){
        this.prevFuelTime = prevFuelTime;
    }

    public long getPrevFuelTime(){
        return this.prevFuelTime;
    }

    public int getFuel(){
        return this.fuelValue;
    }

    // Quang add 10/12/22
    public void setCalibMap(TreeMap<Integer, Integer> calibMap){
        this.calibMap = calibMap;
    }

    // Quang add 10/12/22
    public void calibFuel(int fuelRaw){
        int minLevel = 0;
        int maxLevel = 0;
        int minValue = 0;
        int maxValue = 0;

        Set<Integer> keySet = calibMap.keySet();
        int i = 0;
        Object[] keys = keySet.toArray();
        try{
            if(keys.length > 0){
                minLevel = (Integer)keys[0];
                maxLevel = (Integer)keys[0];
                minValue = calibMap.get(keys[0]);
                maxValue = calibMap.get(keys[0]);
                // cho nay xem lai khi no vuot qua index
//                System.out.println("calib length="+keys.length);
                for(i=0; i<keys.length-1; i++){
                    int keyItem = (Integer)keys[i];
//                    System.out.println("keyItem="+keyItem);
                    if((fuelRaw >= (Integer)keys[i]) && (fuelRaw <= (Integer)keys[i+1])){
                        minLevel = (Integer)keys[i];
                        maxLevel = (Integer)keys[i+1];
                    }
                }

                minValue = calibMap.get(minLevel);
                maxValue = calibMap.get(maxLevel);

                float sampleValue = (float)(maxValue - minValue)/(maxLevel - minLevel);
                int sampleNum = fuelRaw - minLevel;
                float sample = sampleNum * sampleValue;
                fuelValue = (int)(minValue + sample);
            }
        }catch(Exception ex){
            ex.printStackTrace();
            System.out.println("calib error");
        }
    }

    public void setTemp(int tempValue){
        this.tempValue = tempValue;
    }

    public int getTemp(){
        return this.tempValue;
    }

    public void setOverTemp(boolean overTemp){
        this.over_temp = overTemp;
    }

    public boolean isOverTemp(){
        return this.over_temp;
    }

    public void setAircon(boolean airconValue){
        this.aircon_value = airconValue;
    }

    public boolean getAircon(){
        return this.aircon_value;
    }

    public void setTruck(boolean truckValue){
        this.truck_value = truckValue;
    }

    public boolean getTruck(){
        return this.truck_value;
    }

    public void setSD(boolean sd){
        this.sd = sd;
    }

    public boolean getSD(){
        return this.sd;
    }

    public void setDi2(boolean di2){
        this.di2 = di2;
    }

    public boolean getDi2(){
        return this.di2;
    }
    //----------------------------------------------------------------------------------------------------
    // Q add 28/07 : các field sau thuộc bảng device
    public void setPowerId(long power_id){
        this.power_id = power_id;
    }

    public long getPowerId(){
        return this.power_id;
    }

    // Quang add 16/10/22
    public boolean isFixGPS() {
        return this.fixGPS;
    }
    public void setFixGPS(boolean fixGPS){
        this.fixGPS = fixGPS;
    }

    public void setFuelEnable(boolean fuel_enable){
        this.fuelEnable = fuel_enable;
    }

    public boolean isFuelEnable(){
        return this.fuelEnable;
    }

    public void setTankCapac(int tank_capac){
        this.tank_capac = tank_capac;
    }

    public int getTankCapac(){
        return this.tank_capac;
    }

    public void setFuelDelta(int fuel_delta){
        this.fuelDelta = fuel_delta;
    }

    public int getFuelDelta(){
        return this.fuelDelta;
    }

    public void setFuelId(long fuel_id){
        this.fuelId = fuel_id;
    }

    public long getFuelId(){
        return this.fuelId;
    }

    public void setTempEnable(boolean temp_enable){
        this.temp_enable = temp_enable;
    }

    public boolean isTempEnable(){
        return this.temp_enable;
    }

    public void setTempLimit(int temp_limit){
        this.temp_limit = temp_limit;
    }

    public int getTempLimit(){
        return this.temp_limit;
    }

    public void setAirconEnable(boolean aircon_enable){
        this.aircon_enable = aircon_enable;
    }

    public boolean isAirconEnable(){
        return this.aircon_enable;
    }

    public void setAirconId(long aircon_id){
        this.aircon_id = aircon_id;
    }

    public long getAirconId(){
        return this.aircon_id;
    }

    public void setTruckEnable(boolean truck_enable){
        this.truck_enable = truck_enable;
    }

    public boolean isTruckEnable(){
        return this.truck_enable;
    }

    public void setTruckId(long truck_id){
        this.truck_id = truck_id;
    }

    public long getTruckId(){
        return this.truck_id;
    }

    public void setDvrEnable(boolean dvr_enable){
        this.dvr_enable = dvr_enable;
    }

    public boolean isDvrEnable(){
        return this.dvr_enable;
    }

    public void setDvrId(long dvr_id){
        this.dvr_id = dvr_id;
    }

    public long getDvrId(){
        return this.dvr_id;
    }

    public void setCam1Enable(boolean cam1_enable){
        this.cam1_enable = cam1_enable;
    }

    public void setCam2Enable(boolean cam2_enable){
        this.cam2_enable = cam2_enable;
    }

    public void setCam3Enable(boolean cam3_enable){
        this.cam3_enable = cam3_enable;
    }

    public boolean isCam1Enable(){
        return this.cam1_enable;
    }

    public boolean isCam2Enable(){
        return this.cam2_enable;
    }

    public boolean isCam3Enable(){
        return this.cam3_enable;
    }

    public void setPhoto1Path(String photo1Path){
        this.photo1Path = photo1Path;
    }

    public void setPhoto2Path(String photo2Path){
        this.photo2Path = photo2Path;
    }

    public void setPhoto3Path(String photo3Path){
        this.photo3Path = photo3Path;
    }

    public String getPhoto1Path(){
        return this.photo1Path;
    }

    public String getPhoto2Path(){
        return this.photo2Path;
    }

    public String getPhoto3Path(){
        return this.photo3Path;
    }

    // Quang add 29/12/22
    public void setDayMovingElapse(int dayMovingElapse){
        this.dayMovingElapse = dayMovingElapse;
    }

    public int getDayMovingElapse(){
        return this.dayMovingElapse;
    }

    public void setDayStretch(int day_stretch){
        this.day_stretch = day_stretch;
    }

    public int getDayStretch(){
        return this.day_stretch;
    }

    public void setND91(boolean nd91) {
    this.nd91 = nd91;
}

    public boolean isND91() {
        return this.nd91;
    }

    public void setMileageID(long mileageID) {
        this.mileageID = mileageID;
    }

    public long getMileageID() {
        return this.mileageID;
    }

    public void setSpeedLimit(int speed_limit){
        this.speed_limit = speed_limit;
    }

    public int getSpeedLimit(){
        return this.speed_limit;
    }

    public void setDrivingElapse(int drivingElapse){
        this.drivingElapse = drivingElapse;
    }

    public int getDrivingElapse(){
        return this.drivingElapse;
    }

    public void setDrivingOvertime(boolean drivingOvertime){
        this.drivingOvertime = drivingOvertime;
    }

    public boolean isDrivingOvertime(){
        return this.drivingOvertime;
    }

    public void setSpeedVioID(long speedVioID){
        this.speedVioID = speedVioID;
    }

    public long getSpeedVioID(){
        return this.speedVioID;
    }
    //----------------------------------------------------------------------------------------------------
    // các method add sau 01/08/2021 ve trang thai packet
    // Quang add 20/09/22 : taxCode dung de amqp DRVN
    public void setTaxCode(String taxCode){
        this.taxCode = taxCode;
    }

    public String getTaxCode(){
        return this.taxCode;
    }

    // Quang add 04/01/23
    public boolean isLegalTime(){
        // ngay hien tai
        long currentUnixTime = Instant.now().getEpochSecond();
        long delta = Math.abs(currentUnixTime - unixTime);// lech thoi gian
        return (delta < 3*60)? true:false;
    }

    // Quang add 17/02/23
    public void setNumberPlate(String numberPlate){
        this.numberPlate = numberPlate;
    }

    public String getNumberPlate(){
        return this.numberPlate;
    }
}